org.eclipse.scout.rt.mail/src/main/java/org/eclipse/scout/rt/mail/smtp/SmtpConnectionPool.java

```diff
@@ -191,76 +216,82 @@ protected SmtpConnectionPoolEntry leaseConnection(SmtpServerConfig smtpServerCon
           }
         }
       }
-      // we found a valid candidate. Remove it from the idle connection set, add it to the leased connection set and return it
-      m_idleEntries.remove(candidate);
-      m_leasedEntries.add(candidate);
-      LOG.debug("Leasing pooled SMTP connection {}", candidate);
-      return candidate;
+
+      if (createNewConnection) {
+        try {
+          SmtpConnectionPoolEntry candidate = createNewConnectionWithoutPoolLockAndLease(smtpServerConfig);
+          LOG.debug("Leased pooled SMTP connection {}", candidate);
+          return candidate;
+        }
+        catch (AssertionException e) {
+          throw e; // keep AssertionException (do not catch it by catching RuntimeException)
+        }
+        catch (RuntimeException | MessagingException e) {
+          throw new ProcessingException("MessagingException caught while trying to connect to smtp server.", e);
+        }
+        finally {
+          m_currentlyConnectingCounter.compute(smtpServerConfig, (k, v) -> v == 1 ? null : (v - 1)); // decrement (remove if 0)
+        }
+      }
+    }
+  }
+
+  protected void checkAndThrowIfDestroyed() {
```

Comment:
Move to end of class

Reply:
Moved in 231195a99e67f0e19d31639e53a5fd6b7b791174